pkg/cmd/server/origin/asset_apiserver.go

```diff
@@ -0,0 +1,404 @@
+
+type AssetServerConfig struct {
+	GenericConfig *genericapiserver.Config
+
+	Options               oapi.AssetConfig
+	LimitRequestOverrides *clusterresourceoverrideapi.ClusterResourceOverrideConfig
+	IsStandalone          bool
+
+	PublicURL url.URL
+}
+
+// AssetServer serves non-API endpoints for openshift.
+type AssetServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+
+	IsStandalone bool
+	PublicURL    url.URL
+}
+
+type completedAssetServerConfig struct {
+	*AssetServerConfig
+}
+
+// TODO this is taking a very large config for a small piece of it.  The information must be broken up at some point so that
+// we can run this in a pod.  This is an indication of leaky abstraction because it spent too much time in openshift start
+func NewAssetServerConfigFromMasterConfig(masterConfigOptions configapi.MasterConfig) (*AssetServerConfig, error) {
+	assetConfig := masterConfigOptions.AssetConfig
+	publicURL, err := url.Parse(assetConfig.PublicURL)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	_, portString, err := net.SplitHostPort(assetConfig.ServingInfo.BindAddress)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portString)
+	if err != nil {
+		return nil, err
+	}
+	secureServingOptions := apiserveroptions.SecureServingOptions{}
+	secureServingOptions.BindPort = port
+	secureServingOptions.ServerCert.CertKey.CertFile = assetConfig.ServingInfo.ServerCert.CertFile
+	secureServingOptions.ServerCert.CertKey.KeyFile = assetConfig.ServingInfo.ServerCert.KeyFile
+	for _, nc := range assetConfig.ServingInfo.NamedCertificates {
+		sniCert := utilflag.NamedCertKey{
+			CertFile: nc.CertFile,
+			KeyFile:  nc.KeyFile,
+			Names:    nc.Names,
+		}
+		secureServingOptions.SNICertKeys = append(secureServingOptions.SNICertKeys, sniCert)
+	}
+
+	genericConfig := genericapiserver.NewConfig(kapi.Codecs)
+	genericConfig.CorsAllowedOriginList = masterConfigOptions.CORSAllowedOrigins
+	genericConfig.EnableDiscovery = false
+	genericConfig.BuildHandlerChainFunc = buildHandlerChainForAssets(publicURL.Path)
+	if err := secureServingOptions.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	genericConfig.SecureServingInfo.BindAddress = assetConfig.ServingInfo.BindAddress
+	genericConfig.SecureServingInfo.BindNetwork = assetConfig.ServingInfo.BindNetwork
+	genericConfig.SecureServingInfo.MinTLSVersion = crypto.TLSVersionOrDie(assetConfig.ServingInfo.MinTLSVersion)
+	genericConfig.SecureServingInfo.CipherSuites = crypto.CipherSuitesOrDie(assetConfig.ServingInfo.CipherSuites)
+
+	overrideConfig, err := getResourceOverrideConfig(masterConfigOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := &AssetServerConfig{
+		GenericConfig:         genericConfig,
+		Options:               *masterConfigOptions.AssetConfig,
+		LimitRequestOverrides: overrideConfig,
+		PublicURL:             *publicURL,
+		IsStandalone:          masterConfigOptions.AssetConfig.ServingInfo.BindAddress != masterConfigOptions.ServingInfo.BindAddress,
```

Comment:
I hate we do it here and then actual `MasterConfig` has `WebConsoleStandalone` method :/ This asks for some cleanup.